src/string_immutable.zig

```diff
@@ -6678,3 +6745,33 @@ pub fn splitFirstWithExpected(self: string, comptime expected: u8) ?[]const u8 {
     }
     return null;
 }
+
+pub fn percentEncodeWrite(
+    utf8_input: []const u8,
+    writer: *std.ArrayList(u8),
+) !void {
+    var i: usize = 0;
+    while (indexOfNeedsURLEncode(utf8_input[i..])) |j| {
```

Comment:
IMO it's simpler to have a slice that is updated on each run instead of having to [i..] each usage of it